PEP 670: clarify cast; don't change return type (#2349)
python/peps: pep-0670.rst

internal C API exposes implementation details by design and should not be
used outside Python.

Cast pointer arguments
----------------------

Existing cast

> ^^^^^^^^^^^^^

Reviewer: ```suggestion
'''''''''''''
```

Per PEP 12

> Currently, most macros accepting pointers cast pointers arguments to the

Reviewer: ```suggestion
Currently, most macros accepting pointers cast pointer arguments to their
```

Fix grammar error and improve clarity

expected types. For example, in Python 3.6, the ``Py_TYPE()`` macro

> casts its argument to ``PyObject*``::

Reviewer: ```suggestion
casts its argument to ``PyObject*``:

.. code-block:: c
```

To get correct syntax highlighting

    #define Py_TYPE(ob) (((PyObject*)(ob))->ob_type)

> The ``Py_TYPE()`` macro accept ``PyObject*`` type, but also any pointer

Reviewer: ```suggestion
The ``Py_TYPE()`` macro accepts the ``PyObject*`` type, but also any pointer
```

Fix grammar error

Reviewer: Looks like this was fixed

> type like ``PyLongObject*`` or ``PyDictObject*``.

Reviewer: ```suggestion
types, such as ``PyLongObject*`` or ``PyDictObject*``.
```

Fix grammar error and phrasing issue, and add missing comma

Add a new macro to keep the cast

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
''''''''''''''''''''''''''''''''
```

Per PEP 12

When a macro is converted to a function and the macro casts at least one

> of its argument, a new macro is added to keep the cast. The new macro

Reviewer: ```suggestion
of its arguments, a new macro is added to keep the cast. The new macro
```

Grammar error

> and the function have the same name. Example with the ``Py_TYPE()``
> macro converted to a static inline function::

Reviewer: ```suggestion
and the function have the same name. As an example, for the ``Py_TYPE()``
macro converted to a static inline function:

.. code-block:: c
```

Fix grammar errors, phrasing and syntax highlighting

Reviewer: The grammar and phrasing errors weren't fixed here

    static inline PyTypeObject* Py_TYPE(PyObject *ob) {
        return ob->ob_type;
    }
    #define Py_TYPE(ob) Py_TYPE((PyObject*)(ob))

> The cast is kept all pointer types, not only ``PyObject*``.

Reviewer: ```suggestion
The cast is kept for all pointer types, not only ``PyObject*``.
```

Grammar error causing semantic ambiguity; I _believe_ this is what you intended to say.

Removing a cast to ``void*`` would emit a new warning if the function is
called with a variable of ``const void*`` type. For example, the
``PyUnicode_WRITE()`` macro casts its *data* argument to ``void*``, and

> so accept ``const void*`` type, even if it writes into *data*.

Reviewer: ```suggestion
so accepts the ``const void*`` type, even if it writes into *data*.
```

Grammar error

> Avoid the cast in the limited C API 3.11
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
Avoid the cast in the limited C API in 3.11
''''''''''''''''''''''''''''''''''''''''''''
```

Fix grammar error/semantic ambiguity, and follow PEP 12

The cast is removed from the limited C API version 3.11 and newer: the

> caller must pass the expected type, or do the cast. Example with the

Reviewer: ```suggestion
caller must pass the expected type, or perform the cast. An example with the
```

Use appropriate verbiage and fix grammar error

> ``Py_TYPE()`` function::

Reviewer: ```suggestion
``Py_TYPE()`` function:

.. code-block:: c
```

Fix syntax highlighting

    static inline PyTypeObject* Py_TYPE(PyObject *ob) {
        return ob->ob_type;
    }
    #if !defined(Py_LIMITED_API) || Py_LIMITED_API+0 < 0x030b0000
    #  define Py_TYPE(ob) Py_TYPE((PyObject*)(ob))
    #endif

Return type is not changed
--------------------------

When a macro is converted to a function, its return type must not change
to prevent emitting new compiler warnings.

> For example, Python 3.7 changed ``PyUnicode_AsUTF8()`` return type from

Reviewer: ```suggestion
For example, Python 3.7 changed ``PyUnicode_AsUTF8()``\ 's return type from
```

Fix grammar error

``char*`` to ``const char*`` (`commit
<https://github.com/python/cpython/commit/2a404b63d48d73bbaa007d89efb7a01048475acd>`__).
The change emitted new compiler warnings when building C extensions
expecting ``char*``. This PEP doesn't change the return type to prevent
this issue.

Backwards Compatibility
=======================

> The PEP is design to avoid C API incompatible changes.

Reviewer: ```suggestion
The PEP is designed to avoid C-API-incompatible changes.
```

Fix grammar error and correctly hyphenate compound adjective

> Only C extensions targeting explicitly the limited C API version 3.11

Reviewer: ```suggestion
Only C extensions explicitly targeting the limited C API version 3.11
```

Fix phrasing issue (English word order)

  (March 2021).

Version History
===============

> * Version 2: No longer remove return value, remove arguments cast from

Reviewer: ```suggestion
* Version 2: No longer remove return values; remove argument casting from
```

Fix grammar errors (plurality, agreement and comma splice)
